vault/src/metrics.rs

```diff
@@ -0,0 +1,527 @@
+        .get_collateralization_from_vault(vault_id.clone(), false)
+        .await
+        .unwrap_or(0u128);
+    let float_collateralization_percentage = FixedU128::from_inner(collateralization).to_float();
+    metrics.collateralization.set(float_collateralization_percentage);
+
+    let required_collateral = parachain_rpc
+        .get_required_collateral_for_vault(vault_id.clone())
+        .await?;
+    let truncated_required_collateral = FixedU128::from_inner(required_collateral).to_float() * decimals_offset;
+    metrics.required_collateral.set(truncated_required_collateral);
+    Ok(())
+}
+
+pub async fn update_bitcoin_metrics<B: BitcoinCoreApi + Clone + Send + Sync>(
+    vault: VaultData<B>,
+    new_fee_entry: Option<SignedAmount>,
+    fee_budget: Option<u128>,
+) {
+    // update the average fee
+    if let Some(amount) = new_fee_entry {
+        {
+            let mut tmp = vault.metrics.average_btc_fee.data.write().await;
+            *tmp = AverageTracker {
+                total: tmp.total.saturating_add(amount.as_sat().abs() as u64),
+                count: tmp.count.saturating_add(1),
+            };
+        }
+        publish_average_bitcoin_fee(&vault).await;
+
+        if let Ok(budget) = TryInto::<i64>::try_into(fee_budget.unwrap_or(0)) {
+            let surplus = budget.saturating_sub(amount.as_sat().abs());
+            let mut tmp = vault.metrics.fee_budget_surplus.data.write().await;
+            *tmp = tmp.saturating_add(surplus);
+        }
+        publish_fee_budget_surplus(&vault).await;
+    }
+
+    publish_bitcoin_balance(&vault).await;
+}
+
+async fn publish_fee_budget_surplus<B: BitcoinCoreApi + Clone + Send + Sync>(vault: &VaultData<B>) {
+    let surplus = *vault.metrics.fee_budget_surplus.data.read().await;
+    vault
+        .metrics
+        .fee_budget_surplus
+        .gauge
+        .set(surplus as f64 / vault.vault_id.wrapped_currency().one() as f64);
+}
+
+async fn publish_average_bitcoin_fee<B: BitcoinCoreApi + Clone + Send + Sync>(vault: &VaultData<B>) {
+    let average = match vault.metrics.average_btc_fee.data.read().await {
+        x if x.count > 0 => x.total as f64 / x.count as f64,
+        _ => 0.0,
+    };
+    vault.metrics.average_btc_fee.gauge.set(average);
+}
+
+async fn publish_bitcoin_balance<B: BitcoinCoreApi + Clone + Send + Sync>(vault: &VaultData<B>) {
+    match vault.btc_rpc.get_balance(None).await {
+        Ok(bitcoin_balance) => vault.metrics.btc_balance.actual.set(bitcoin_balance.as_btc() as f64),
+        Err(e) => {
+            // unexpected error, but not critical so just continue
+            tracing::warn!("Failed to get Bitcoin balance: {}", e);
+        }
+    }
+}
+
+async fn publish_kint_balance(parachain_rpc: &InterBtcParachain) {
+    if let Ok(balance) = parachain_rpc.get_free_balance(parachain_rpc.native_currency_id).await {
+        KINT_BALANCE.set(balance as f64 / parachain_rpc.native_currency_id.one() as f64);
```

Comment:
Really? Wouldn't that make things more difficult? Since e.g. intr and kint don't have the same number of decimals

Reply:
In this case the BTC balance and fees should also be in Satoshi. Grafana seems to allow setting custom Y-axis units https://grafana.com/docs/grafana/latest/visualizations/time-series/change-axis-display/

Reply:
Because it shifts the burden of making the conversions to the vault operator - they will need to know and modify the number of decimals for each currency they use. I don't really understand why you want to take away the automation and rely on the operators instead - what is the advantage?